lib/utils/seqid/src/test/test.js

```diff
@@ -20,57 +17,96 @@ describe('abacus-seqid', () => {
+        // id to be greater than the previous one
+        clock.tick(100);
+        const id3 = seqid();
+        expect(id3).to.match(/-0-0-0-0/);
+        expect(id3 > id2).to.be.true;
+
+        // Get an id for an earlier time (simulating clock skew), expect the
+        // id time to still be the later time and the id counter to increase
+        // to make the id greater than the previous one
+        clock.tick(-200);
+        const id4 = seqid();
+        expect(id4).to.match(/-0-0-0-1/);
+        expect(id4 > id3).to.be.true;
+      });
+    });
+
+    context('with timestamp', () => {
+      it('uses the timestamp for sequential ids', () => {
+        expect(seqid(1537596306000)).to.match(/1537596306000-\d{16}-0-0-0-0/);
+        expect(seqid(1537596306000)).to.match(/1537596306000-\d{16}-0-0-0-1/);
+      });
+
+      it('throws on too long timestamp', () => {
+        const sequenceFn = () => seqid('1234567891011121314');
+        expect(sequenceFn).to.throw('Not a timestamp');
+      });
+
+      it('throws on too short timestamp', () => {
+        const sequenceFn = () => seqid('123');
+        expect(sequenceFn).to.throw('Not a timestamp');
+      });
+
+      it('throws on bad characters in timestamp', () => {
+        const sequenceFn = () => seqid('15375@#%06000');
+        expect(sequenceFn).to.throw('Not a timestamp');
+      });
+
+    });
   });
 
-  it('samples unique sequential ids', () => {
-    const id = seqid();
-    const t = parseInt(id);
-    expect(seqid.sample(undefined)).to.equal(undefined);
-    expect(seqid.sample(id)).to.equal(id);
-    expect(seqid.sample(id, 1)).to.equal('000' + t);
-    expect(seqid.sample(id, 3600000)).to.equal('000' + Math.floor(t / 3600000) * 3600000);
-    expect(seqid.sample(id, '180s')).to.equal('000' + Math.floor(t / 180000) * 180000);
-    expect(seqid.sample(id, 's')).to.equal('000' + Math.floor(t / 1000) * 1000);
-    expect(seqid.sample(id, '15m')).to.equal('000' + Math.floor(t / 900000) * 900000);
-    expect(seqid.sample(id, 'm')).to.equal('000' + Math.floor(t / 60000) * 60000);
-    expect(seqid.sample(id, '2h')).to.equal('000' + Math.floor(t / 7200000) * 7200000);
-    expect(seqid.sample(id, 'h')).to.equal('000' + Math.floor(t / 3600000) * 3600000);
-    expect(seqid.sample(id, '5D')).to.equal('000' + Math.floor(t / 432000000) * 432000000);
-    expect(seqid.sample(id, 'D')).to.equal('000' + Math.floor(t / 86400000) * 86400000);
-    expect(seqid.sample(id, '2M')).to.equal('000' + Date.UTC(2016, 8));
-    expect(seqid.sample(id, 'M')).to.equal('000' + Date.UTC(2016, 9));
-    const fn = () => seqid.sample(id, 'Y');
-    expect(fn).to.throw();
+  describe('sample', () => {
+
+    it('samples unique sequential ids', () => {
+      const id = seqid();
+      const t = parseInt(id);
+      expect(seqid.sample(undefined)).to.equal(undefined);
+      expect(seqid.sample(id)).to.equal(id);
+      expect(seqid.sample(id, 1)).to.equal('000' + t);
```

Comment:
We may consider to extract '000' in constant (commonPrefix or something more meaningful) since it is used a lot.